libs/libc/machine/risc-v/arch_elf.c

```diff
@@ -317,58 +396,101 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
         break;
 
       case R_RISCV_PCREL_LO12_I:
-      case R_RISCV_PCREL_LO12_S:
         {
+          long imm_hi;
+          long imm_lo;
+
           binfo("%s at %08" PRIxPTR " [%08" PRIx32 "] "
                 "to sym=%p st_value=%08lx\n",
                 _get_rname(relotype),
                 addr, _get_val((uint16_t *)addr),
                 sym, sym->st_value);
 
-          /* NOTE: imm value for mv has been adjusted in previous HI20 */
+          offset = _find_hi20(arch_data, sym->st_value);
+
+          /* Adjust imm for MV(ADDI) / JR (JALR) : I-type */
+
+          _calc_imm(offset, &imm_hi, &imm_lo);
+
+          _add_val((uint16_t *)addr, (int32_t)imm_lo << 20);
         }
         break;
 
-      case R_RISCV_PCREL_HI20:
-      case R_RISCV_CALL:
-      case R_RISCV_CALL_PLT:
+      case R_RISCV_PCREL_LO12_S:
         {
+          uint32_t val;
+          long imm_hi;
+          long imm_lo;
+
           binfo("%s at %08" PRIxPTR " [%08" PRIx32 "] "
                 "to sym=%p st_value=%08lx\n",
                 _get_rname(relotype),
                 addr, _get_val((uint16_t *)addr),
                 sym, sym->st_value);
 
-          offset = (long)sym->st_value + (long)rel->r_addend - (long)addr;
+          offset = _find_hi20(arch_data, sym->st_value);
+
+          /* Adjust imm for SW : S-type */
+
+          _calc_imm(offset, &imm_hi, &imm_lo);
+
+          val = (((int32_t)imm_lo >> 5) << 25) +
+                (((int32_t)imm_lo & 0x1f) << 7);
 
+          binfo("imm_lo=%ld (%lx), val=%" PRIx32 "\n", imm_lo, imm_lo, val);
+
+          _add_val((uint16_t *)addr, val);
+        }
+        break;
+
+      case R_RISCV_PCREL_HI20:
+        {
           long imm_hi;
           long imm_lo;
 
+          binfo("%s at %08" PRIxPTR " [%08" PRIx32 "] "
+                "to sym=%p st_value=%08lx\n",
+                _get_rname(relotype),
+                addr, _get_val((uint16_t *)addr),
+                sym, sym->st_value);
+
+          offset = (long)sym->st_value + (long)rel->r_addend - (long)addr;
+
           _calc_imm(offset, &imm_hi, &imm_lo);
 
           /* Adjust auipc (add upper immediate to pc) : 20bit */
 
-          _add_val((uint16_t *)addr, (imm_hi << 12));
+          _add_val((uint16_t *)addr, imm_hi << 12);
```

Comment:
LGTM! But in the light of the my recent investigations I wonder if we should assert or return an error in case of RV64 and `imm_hi` having a sign bit set?

Reply:
Reference: https://github.com/riscv-collab/riscv-gnu-toolchain/commit/a36a4fba58a3f04e9fa439c16e8d6e453ef6067d

Reply:
If yes, then we need to update `R_RISCV_HI20` as well

Reply:
I think this merits further investigation, yes. It seems it depends on the code model https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc

The affected type is medlow, medany should work (I / we use medany for RV64 AFAIK) ? So this error should not affect us if I understand correctly.

Regardless, not related to this PR.

Reply:
No. Please move forward with merge.
@pussuw Both `medlow` (https://github.com/riscv-collab/riscv-gnu-toolchain/issues/103) and `medany` (https://github.com/riscv-collab/riscv-gnu-toolchain/issues/733) modes are affected on RV64. Only RV32 is not affected as there is not place for sign bit extension obviously. The `medlow` is affected more compared to `medany` because of its nature.
The https://github.com/riscv-collab/riscv-gnu-toolchain/issues/733 describes clearly how this can be detected by linker. From my investigations it is enough to have bit 31 set in symbol address to get linker error, so it highly depends on the SoC address mapping.
I would not argue if this is or is not related to this PR, but rather spotted things I've met recently, so please feel free to make a merge decision.

Reply:
Yes thanks @pkarashchenko. I think I described what I meant poorly.

auipc + whatever pair works, the issue is with lui + whatever. When compiling C-code with medany the compiler should not produce lui + whatever pairs AFAIK. What the person in #733 did was write an assembly routine with lui + hi/prcel_hi, which is not medany.

The reason I have not encountered such issues ever is that I never write the lui/auipc instructions directly, instead I use macro instructions (like `la`/ `call`/ `tail`) which produce the correct auipc + jr result when using medany.

And yes, now that you pointed out the other topic I do agree with you; we need to check for msbit in all _HI20 relocation types (for RV64) to catch this type of usage. The linker will still produce an error though.

I think returning an error code might be better (process spawn fails with errorcode), instead of assertion (coredump), what do you think ?

Reply:
I suggest we merge this patch, it is not related to this sign extension stuff, it just fixes the linking of pcrel_hi to obey what the risc-v spec. says

Reply:
Yes, returning error seems to be reasonable. Lets merge the patch and the rest can be improved as a next step.
I met similar error when I was working with SoC that had SRAM mapped starting from `0xb0000000` address while my initial loader code was executing from XIP mapped starting from `0x3000000`, so I had a linker defined symbol defined as a jump point for the next level SW that was `0xb0000XXX`, so as soon as I used that linker defined symbol in my C code (that was executed from XIP) to cast it to function pointer and make a jump the linker was generating a relocation error even with `medany` variant. I started my experiments and did many attempts in order to overcome this issue.
I would disagree with your statement about assembly code no relation to `medany` as https://github.com/riscv-collab/riscv-gnu-toolchain/issues/733 description clearly states used cmodel `// in c, compile with -mcmodel=medany`. I think that is exactly to show that issue is not `medlow` specific. Usually compiler will generate a PC relative instructions and that I agree, but I also tried to overcome issue in my case (that by default was `auipc`+pair) by changing it to inline assembly of `lui`+pair and both variants lead to relocation error at link time because both `auipc` and `lui` have the same nature of sign bit propagation with `prcel_hi`/`hi`.

Reply:
Yes, it might be that I'm missing something, not having had to debug this issue myself :)

Regardless, we must patch the HI20 relocations, because user can manually write a lui instruction that uses a symbol with the msbit set. I can do a patch for it, should be simple enough to test as well (with the instructions from 733).